### PR TITLE
Support for volumes in K8S/Openshift type components

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
Support for volumes in K8S/Openshift type components

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/15122

### CODE REVIEW

**Comment by dmytro-ndp** [comment]
crw-ci-test

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplierTest.java [hunk 1]
```diff
@@ -266,6 +269,85 @@ public void shouldProvisionProjectVolumesIfSpecifiedIntoK8SList() throws Excepti
...
+
+    // when
+    applier.apply(workspaceConfig, component, s -> yamlRecipeContent);
+  }
+
+  @Test(
+      expectedExceptions = DevfileException.class,
+      expectedExceptionsMessageRegExp =
+          "Conflicting volume with same path \\('/foo/bar'\\) but different name \\('foo'\\) found for component 'foo' and its container 'server'.")
+  public void shouldThrowExceptionWhenDevfileVolumePathIsExists() throws Exception {
```
Comment (metlos) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplierTest.java]: `....PathExists()`

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplierTest.java [hunk 2]
```diff
@@ -266,6 +269,85 @@ public void shouldProvisionProjectVolumesIfSpecifiedIntoK8SList() throws Excepti
...
+                  entry ->
+                      entry.getKey().equals("bar") && entry.getValue().getPath().equals("/bar1")));
+    }
+  }
+
+  @Test(
+      expectedExceptions = DevfileException.class,
+      expectedExceptionsMessageRegExp =
+          "Conflicting volume with same name \\('foo_volume'\\) but different path \\('/foo1'\\) found for component 'foo' and its container 'server'.")
+  public void shouldThrowExceptionWhenDevfileVolumeNameIsExists() throws Exception {
```
Comment (metlos) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplierTest.java]: `...NameExists()`

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/devfile/validator/DevfileSchemaValidatorTest.java [hunk 3]
```diff
@@ -50,7 +50,7 @@ public void shouldNotThrowExceptionOnValidationOfValidDevfile(String resourceFil
       {"editor_plugin_component/devfile_editor_component_with_custom_registry.yaml"},
       {"editor_plugin_component/devfile_editor_plugins_components_with_resource_limits.yaml"},
       {"editor_plugin_component/devfile_plugin_components_with_preferences.yaml"},
-      {"kubernetes_openshift_component/devfile_kubernetes_component_reference.yaml"},
```
Comment (metlos) [inline wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/devfile/validator/DevfileSchemaValidatorTest.java:53]: Is `devfile_kubernetes_component_reference.yaml` used anywhere?

**Review by metlos — COMMENTED** [review]
LGTM, just a few minor nitpicks...

**Review by sparkoo — APPROVED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java [hunk 4]
```diff
@@ -215,23 +215,62 @@ private void applyProjectsVolumes(List<PodData> podsData, List<HasMetadata> comp
...
         machineConfigs.put(machineName, config);
       }
     }
     return machineConfigs;
   }
 
+  private void provisionVolumes(
+      ComponentImpl component, Container container, MachineConfigImpl config)
+      throws DevfileException {
+    for (org.eclipse.che.api.workspace.server.model.impl.devfile.VolumeImpl v :
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java]: can we rename the variable to for example `componentVolume`?

Comment (sparkoo) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java]: I would prefer to use more descriptive names in bigger blocks, and shortcuts for lambdas

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java [hunk 5]
```diff
@@ -215,23 +215,62 @@ private void applyProjectsVolumes(List<PodData> podsData, List<HasMetadata> comp
...
+        component.getVolumes()) {
+      Optional<VolumeMount> sameNameMount =
+          container
+              .getVolumeMounts()
+              .stream()
+              .filter(volume -> volume.getName().equals(v.getName()))
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java]: rename `volume` to `volumeMount` or `vm` ?

**Review by sparkoo — APPROVED** [review]
just few ideas for var names

File: wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/devfile/validator/DevfileSchemaValidatorTest.java (hunk 3 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-workspace/src/test/java/org/eclipse/che/api/workspace/server/devfile/validator/DevfileSchemaValidatorTest.java:53]: just file rename

**Comment by mshaposhnik** [comment]
[ci-test]

**Review by metlos — APPROVED** [review]

**Review by amisevsk — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]
Docs pr?

**Pull request merged**